types/block.go

```diff
@@ -68,3 +74,166 @@ type Signature []byte
+	h.Version.Block = other.Version.Block
+	h.Version.App = other.Version.App
+	if !safeCopy(h.NamespaceID[:], other.NamespaceId) {
+		return errors.New("invalid lenght of 'NamespaceId'")
+	}
+	h.Height = other.Height
+	h.Time = other.Time
+	if !safeCopy(h.LastHeaderHash[:], other.LastHeaderHash) {
+		return errors.New("invalid lenght of 'InvalidLastHeaderHash'")
+	}
+	if !safeCopy(h.LastCommitHash[:], other.LastCommitHash) {
+		return errors.New("invalid lenght of 'InvalidLastCommitHash'")
+	}
+	if !safeCopy(h.DataHash[:], other.DataHash) {
+		return errors.New("invalid lenght of 'InvalidDataHash'")
+	}
+	if !safeCopy(h.ConsensusHash[:], other.ConsensusHash) {
+		return errors.New("invalid lenght of 'InvalidConsensusHash'")
+	}
+	if !safeCopy(h.AppHash[:], other.AppHash) {
+		return errors.New("invalid lenght of 'InvalidAppHash'")
+	}
+	if !safeCopy(h.LastResultsHash[:], other.LastResultsHash) {
+		return errors.New("invalid lenght of 'InvalidLastResultsHash'")
+	}
+	if len(other.ProposerAddress) > 0 {
+		h.ProposerAddress = make([]byte, len(other.ProposerAddress))
+		copy(h.ProposerAddress, other.ProposerAddress)
+	}
+
+	return nil
+}
+
+// safeCopy copies bytes from src slice into dst slice if both have same size.
+// It returns true if sizes of src and dst are the same.
+func safeCopy(dst, src []byte) bool {
+	if len(src) != len(dst) {
+		return false
+	}
+	_ = copy(dst, src)
+	return true
+}
+
+func (b *Block) ToProto() *pb.Block {
+	return &pb.Block{
+		Header:     b.Header.ToProto(),
+		Data:       b.Data.ToProto(),
+		LastCommit: b.LastCommit.ToProto(),
+	}
+}
+
+func (d *Data) ToProto() *pb.Data {
+	return &pb.Data{
+		Txs:                    txsToByteSlices(d.Txs),
+		IntermediateStateRoots: d.IntermediateStateRoots.RawRootsList,
+		Evidence:               evidenceToProto(d.Evidence),
+	}
+}
+
+func (c *Commit) ToProto() *pb.Commit {
+	return &pb.Commit{
+		Height:     c.Height,
+		HeaderHash: c.HeaderHash[:],
+		Signatures: signaturesToByteSlices(c.Signatures),
+	}
+}
+
+func (b *Block) FromProto(other *pb.Block) error {
+	err := b.Header.FromProto(other.Header)
+	if err != nil {
+		return err
+	}
+	b.Data.Txs = byteSlicesToTxs(other.Data.Txs)
+	b.Data.IntermediateStateRoots.RawRootsList = other.Data.IntermediateStateRoots
+	b.Data.Evidence = evidenceFromProto(other.Data.Evidence)
+	if other.LastCommit != nil {
+		b.LastCommit.Height = other.LastCommit.Height
+		if !safeCopy(b.LastCommit.HeaderHash[:], other.LastCommit.HeaderHash) {
+			return errors.New("invalid lenght of 'InvalidLastCommit.HeaderHash'")
+		}
+		b.LastCommit.Signatures = byteSlicesToSignatures(other.LastCommit.Signatures)
+	}
+
+	return nil
+}
+
+func txsToByteSlices(txs Txs) [][]byte {
+	bytes := make([][]byte, len(txs))
+	for i := range txs {
+		bytes[i] = txs[i]
+	}
+	return bytes
+}
+
+func byteSlicesToTxs(bytes [][]byte) Txs {
+	if len(bytes) == 0 {
+		return nil
+	}
+	txs := make(Txs, len(bytes))
+	for i := range txs {
+		txs[i] = bytes[i]
```

Comment:
Is there a reason to assign here instead of copy? Does this have strange consequences if the underlying data is modified with a reference to `bytes` instead of the `txs`?

Reply:
Indeed, it will have some strange consequences if someone decides to change Proto struct after deserialization(look at `byteSlicesToTxs` usage). However, it is not practical to do a copy every time for this unrealistic case, IMO. Also, both Tendermint and IPFS do not do copy in such cases.

Reply:
From a philosophical perspective, I see it as: all else being equal you should prefer safety over performance. If Rust were used, then it could provide zero-cost compile-time type safety to guarantee immutability of the backing arrays. Since we're using Go and we don't get these guarantees, taking the performance hit and doing a copy is better than trying to save an unknown amount of performance (how much is really saved???) and making the code harder to reason about.

Anyways, that's just me ranting. Feel free to ignore.

Reply:
The programming languages, as you know, develop some level of mindset for their devs. From Rust's perspective, I agree with you, though I didn't write more than ten loc on it, but for Go, that's slightly different. 

In Go, a package maintainer can do everything he wants inside his package, even potentially unsafe, but his duty is to provide safe API to those who import the package. That's how to Go’s std, and its internal are written. There are many hacks in it, that later even got reused by Gophers in their projects. 

Usually, I hide serialization helper struts, etc., without giving access from outside. This way, I can ensure that the API is safe and a user won't change anything by accident. The problem though still exists for package devs, as they can forget about such things in their code and consequently shoot their legs through time... Not sure if Rust is a savior here, I am looking forward to Zig.